### DISCUSSION TITLE
Is charm deterministic -- will the same T1/T2 images always create the same head model?

### DISCUSSION THREAD

**Discussion by grodgers1** [simnibs/simnibs#304] [7.Creating head models]
I ran charm on the same T1/T2 images (different filenames, but exactly the same data inside the files) and find some very tiny differences in the resulting head models. For example, the nonlinear transformations change by a tiny amount, the number of nodes in the mesh changes by about 0.1%, etc. It looks to me like maybe the non-linear registration is changing between runs, which in turn causes minor changes in segmentation and thus meshing. What part of the registration would cause different results for the same inputs? Does this mean that charm is not deterministic?

**Comment by oulap** [comment]
Hey, it seems there are small differences between runs as you noticed. We are tracking down the exact place where this happens. I'll ping this discussion again once we find where the differences are coming from.

**Comment by merdi-dtu** [comment]
Thank you for pointing this out. We found a stochastic component in the registration of the T1 and T2 images, which we fixed and it will go out with the next release. The meshing however remains with some non-deterministic components, which is independent of the deformation part. The meshing stochasticity comes from cgal where we do not have direct influence on.
